tests/test_database.py

```diff
@@ -167,3 +168,16 @@ def test_default_precision(self):
         raw_schema = db.query_table_schema(t.path)
         schema = db._process_table_schema(t.path, raw_schema)
         self.assertEqual(schema["value"].precision, db.dialect.DEFAULT_NUMERIC_PRECISION)
+
+
+class TestDatabaseConnection(unittest.TestCase):
+    def test_close_connection_postgres(self):
+        connection: Database = get_conn(dbs.PostgreSQL)
+        with connection:
+            cursor = connection._conn.cursor()
+            cursor.execute("SELECT 1")
+
+        with self.assertRaises(InterfaceError) as interface_error:
+            cursor.execute("SELECT 1")
+
+        assert str(interface_error.exception) == "cursor already closed"
```

Comment:
```suggestion
@test_each_database
class TestCloseMethod(unittest.TestCase):
    def test_close_connection(self):
        connection: Database = get_conn(self.db_cls)
        
        # Perform a query to verify the connection is established
        with connection:
            cursor = connection._conn.cursor()
            cursor.execute("SELECT 1")

        # Now the connection should be closed, and trying to execute a query should fail.
        with self.assertRaises(Exception):  # Catch any type of exception.
            cursor.execute("SELECT 1")
            
```

Reply:
This tests every database with the `@test_each_database` decorator. 

Updated `mysql.py` to get this passing. You'll have to update your branch for the below. 

```python
@attrs.define(frozen=False, init=False, kw_only=True)
class MySQL(ThreadedDatabase):
    DIALECT_CLASS: ClassVar[Type[BaseDialect]] = Dialect
    SUPPORTS_ALPHANUMS = False
    SUPPORTS_UNIQUE_CONSTAINT = True
    CONNECT_URI_HELP = "mysql://<user>:<password>@<host>/<database>"
    CONNECT_URI_PARAMS = ["database?"]

    _args: Dict[str, Any]
    _conn: Any

    def __init__(self, *, thread_count, **kw) -> None:
        super().__init__(thread_count=thread_count)
        self._args = kw

        # In MySQL schema and database are synonymous
        try:
            self.default_schema = kw["database"]
        except KeyError:
            raise ValueError("MySQL URL must specify a database")

    def create_connection(self):
        mysql = import_mysql()
        try:
            self._conn = mysql.connect(charset="utf8", use_unicode=True, **self._args)
            return self._conn
        except mysql.Error as e:
            if e.errno == mysql.errorcode.ER_ACCESS_DENIED_ERROR:
                raise ConnectError("Bad user name or password") from e
            elif e.errno == mysql.errorcode.ER_BAD_DB_ERROR:
                raise ConnectError("Database does not exist") from e
            raise ConnectError(*e.args) from e
```

Reply:
> 

For this close method test to pass for all databases, you'll have to update `vertica.py` and `presto.py`

based on the changes I made in this mirror branch.

Reply:
@sungchun12 we do not have to close `vertica.py` since we create connections for each thread and these connections are being closed in the parent.

for `Presto` I have added the required changes.

Reply:
@sar009 Looks like we do have to call it. I just removed the `close` method override in `vertica.py` and it failed in my mirror PR: https://github.com/datafold/data-diff/actions/runs/7534996859/job/20510263941?pr=841#step:8:75

https://github.com/datafold/data-diff/pull/841/commits/a0666de88528670d49d95782588dd4d3b1823819

The test fails because you haven't defined a ﻿`close()` method in the ﻿`Vertica` class to close the individual connection(﻿`self._conn`).

Even though you mentioned that the ﻿`close()` method should be inherited from ﻿`ThreadedDatabase`, that's not the case in this application because the ﻿`close()` method provided in ﻿`ThreadedDatabase` performs tasks that consist closing all threads and the connection in the ﻿`thread_local`. However, it will not handle closing of the individual connection (`﻿self._conn`) which is initiated in the method ﻿`create_connection()` of the ﻿`Vertica` class.

Reply:
You resolved this but didn't fix this issue. Opening it back up.

Reply:
@sungchun12 the `create_connection` in `vertica.py` should not set class attribute connection(`self._conn`) rather like all other classes `vertica.py` should as well return a new instance in `create_connection`, as the method `create_connection` in `ThreadedDatabase` is called from `set_conn` in `ThreadedDatabase` and the db connection is managed in `self.thread_local` of `ThreadedDatabase`. The connection set in this fashion is closed from close method of `ThreadedDatabase`.

Your pr is failing after commenting the close function in `vertica.py` because the connection is used from `self._conn` and not from `self.thread_local.conn` of `ThreadedDatabase`. If you revert your change in `create_connection` of `vertica.py` to original as follow the build should be successful.
```python
    def create_connection(self):
        vertica = import_vertica()
        try:
            c = vertica.connect(**self._args)
            return c
        except vertica.errors.ConnectionError as e:
            raise ConnectError(*e.args) from e
```

P.S. only Postgres' behavior is different I will take a look at that as well once this PR is merged.

Reply:
@sar009 I reverted to the original and it fails the test both with and without the `_conn` attribute added. 

https://github.com/datafold/data-diff/actions/runs/7587669203/job/20668531425?pr=841
https://github.com/datafold/data-diff/actions/runs/7587791525/job/20668911798?pr=841

Unless there's something else I'm missing, I recommend you follow my code suggestion!

After you make the update, let me know when this is ready for a full review. We're almost at the finish line!

Reply:
@sungchun12 Thanks You, the tests are failing because additionally what I have posted above the test method in `test_database.py` should look something like this

```
@test_closeable_databases
class TestCloseMethod(unittest.TestCase):
    def test_close_connection(self):
        database: Database = get_conn(self.db_cls)

        # Perform a query to verify the connection is established
        with database:
            database.query("SELECT 1")

        # Now the connection should be closed, and trying to execute a query should fail.
        with self.assertRaises(Exception):  # Catch any type of exception.
            database.query("SELECT 1")
```

in your code you are accessing the `_conn` object explicitly. Let the db class decide how to manage the connection. Also as with all the code I am recommending is already in this PR and the tests are executing fine.

Reply:
Thanks for merging the PR @sungchun12